provider/uprops/src/enumerated.rs

```diff
@@ -0,0 +1,306 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::error::Error;
+use crate::upropdump_serde;
+use icu_provider::iter::IterableDataProviderCore;
+use icu_provider::prelude::*;
+use icu_uniset::provider::*;
+use icu_uniset::UnicodeSetBuilder;
+use std::borrow::Cow;
+use std::fs;
+use std::path::PathBuf;
+
+pub struct EnumeratedPropertiesDataProvider {
+    root_dir: PathBuf,
+}
+
+/// A data provider reading from .toml files produced by the ICU4C upropdump tool.
+impl EnumeratedPropertiesDataProvider {
+    pub fn new(root_dir: PathBuf) -> Self {
+        EnumeratedPropertiesDataProvider { root_dir }
+    }
+    fn get_toml_data(&self, name: &str) -> Result<upropdump_serde::enumerated::Main, Error> {
+        let mut path: PathBuf = self.root_dir.clone().join(name);
+        path.set_extension("toml");
+        let toml_str = fs::read_to_string(&path).map_err(|e| Error::Io(e, path.clone()))?;
+        toml::from_str(&toml_str).map_err(|e| Error::Toml(e, path))
+    }
+}
+
+fn expand_groupings<'a>(prop_name: &str, prop_val: &'a str) -> Vec<&'a str> {
+    match (prop_name, prop_val) {
+        // General_Category::CasedLetter
+        ("gc", "LC") => vec!["Lu", "Ll", "Lt"],
+
+        // General_Category::Letter
+        ("gc", "L") => vec!["Lu", "Ll", "Lt", "Lm", "Lo"],
+
+        // General_Category::Mark
+        ("gc", "M") => vec!["Mn", "Mc", "Me"],
+
+        // General_Category::Number
+        ("gc", "N") => vec!["Nd", "Nl", "No"],
+
+        // General_Category::Punctuation
+        ("gc", "P") => vec!["Pc", "Pd", "Ps", "Pe", "Pi", "Pf", "Po"],
+
+        // General_Category::Symbol
+        ("gc", "S") => vec!["Sm", "Sc", "Sk", "So"],
+
+        // General_Category::Separator
+        ("gc", "Z") => vec!["Zs", "Zl", "Zp"],
+
+        // General_Category::Control
+        ("gc", "C") => vec!["Cc", "Cf", "Cs", "Co", "Cn"],
+
+        (_, prop_val) => vec![prop_val],
+    }
```

Comment:
`minor, suggestion (n-b)`:

```rust
match prop_name {
    "gc" => match prop_val {
        "LC" =>...,
        "L" => ...,
    },
    _ => vec![prop_val],
}
```

would lead to a slightly lower cognitive overhead for the reader.

Reply:
Good idea.